File: src/main/tsx/types/Category.ts
import TypeMapper from "@/mapper/TypeMapper";
import CategorySmoothType from "@/types/CategorySmoothType";
import Target, {TargetDTO, targetMapper} from "@/types/Target";
import TransactionType from "@/types/TransactionType";

interface Category {
    id: number;
    parentId: number | null;
    label: string;
    transactionType: TransactionType;
    smoothType: CategorySmoothType;
    targets: Array<Target>;
}

export interface CategoryDTO {
    id: number;
    parentId: number | null;
    label: string;
    transactionType: TransactionType;
    smoothType: CategorySmoothType;
    targets: Array<TargetDTO>;
}

export const categoryMapper: TypeMapper<Category, CategoryDTO> = {
    fromDTO: (dto: CategoryDTO) => ({
        ...dto,
        targets: dto.targets.map(targetMapper.fromDTO),
    }),
    toDTO: (model: Category) => ({
        ...model,
        targets: model.targets.map(targetMapper.toDTO),
    }),
};

export default Category;

File: backend/src/main/java/com/lennartmoeller/finance/model/Category.java
package com.lennartmoeller.finance.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import java.util.List;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.RequiredArgsConstructor;

@Data
@Entity
@EqualsAndHashCode(of = "id", callSuper = false)
@RequiredArgsConstructor
@Table(name = "categories")
public class Category extends BaseModel {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent")
    private Category parent;

    @Column(nullable = false, unique = true)
    private String label;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionType transactionType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CategorySmoothType smoothType = CategorySmoothType.DAILY;

    @Column
    private String icon;

    // TODO: Maybe add domain name to fetch logo from https://docs.logo.dev

    @OneToMany(mappedBy = "category", fetch = FetchType.LAZY)
    private List<Target> targets;
}
